lib/pbench/server/api/resources/query_apis/datasets/table_of_content.py

```diff
@@ -0,0 +1,210 @@
+from logging import Logger
+
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    JSON,
+    Schema,
+    Parameter,
+    ParamType,
+)
+from pbench.server.api.resources.query_apis import CONTEXT
+from pbench.server.api.resources.query_apis.datasets import RunIdBase
+
+
+class TableOfContent(RunIdBase):
+    """
+    Table of Content API returns the list of sub-directories and files
+    present under a directory.
+    """
+
+    MAX_SIZE = 10000
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("run_id", ParamType.STRING, required=True),
+                Parameter("parent", ParamType.STRING, required=True),
+            ),
+        )
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        """
+        Construct a pbench Elasticsearch query for getting a list of
+        documents which contains the user provided parent with files
+        and its sub-directories with metadata of run-toc index document
+        that belong to the given run id.
+
+        Args:
+            json_data: JSON dictionary of type-normalized parameters,
+                    we pull 'parent' key value from it.
+            context: propagate the requested set of metadata to other
+                    methods, it includes "run_id" which is Dataset
+                    document ID.
+
+        EXAMPLE:
+        {
+            "run_id": "1234567890"
+            "parent": '/1-default'
+        }
+        """
+        # Copy parent directory metadata to CONTEXT for postprocessor
+        parent = context["parent"] = json_data.get("parent")
+        run_id = context["run_id"]
+        dataset = context["dataset"]
+
+        self.logger.info(
+            "Discover Table of Content run_data_id {}, directory {}",
+            dataset.md5,
+            parent,
+        )
+
+        # Retrieve the ES indices that belong to this run_id from the metadata
+        # table
+
+        indices = self.get_index(dataset, "run-toc")
+        if not indices:
+            self.logger.debug(
+                f"Found no indices matching the prefix run-toc"
+                f" for a dataset {dataset!r}"
+            )
+            abort(HTTPStatus.NOT_FOUND, message="Found no matching indices")
+
+        return {
+            "path": f"/{indices}/_search",
+            "kwargs": {
+                "json": {
+                    "size": self.MAX_SIZE,
+                    "query": {
+                        "bool": {
+                            "filter": [
+                                {
+                                    "dis_max": {
+                                        "queries": [
+                                            {"match_phrase": {"directory": parent}},
+                                            {"match_phrase": {"parent": parent}},
```

Comment:
`"parent"` is a keyword type. I'm not sure of the full technical Elasticsearch distinction here, but perhaps `"match"` would be more efficient? (I don't think we need/want an analyzer for parent.)

Only `"directory"` and `{"files": {"linkpath"}}` are path-analyzed `"text"` fields.

Reply:
My understanding is that _keyword_ fields are dealt with as a single term, while _text_ fields are **_analyzed_** (which, by default, means they are broken up into a list of terms, but, in our case, means they have been deconstructed as a series of file paths), so on a keyword field you can do a `"term"` query (which is simple, fast, and easily understood), but you cannot do the same on a text field (unless it has a [multi-field](https://www.elastic.co/guide/en/elasticsearch/reference/7.13/multi-fields.html) mapping, which has both a _keyword_ and a _text_ representation...but we cannot change the field mappings without re-indexing); instead, a text field requires a `"match"` query (or some other more complicated query).

I believe that `"match"` (and `"match_phrase"` -- I'm not sure what the difference between the two is...or if there even _is_ a **functional** difference -- [the documentation for `"match_phrase"`](https://www.elastic.co/guide/en/elasticsearch/reference/7.13/query-dsl-match-query-phrase.html) is poor even by Elasticsearch standards!) is a reasonably inexpensive query, so it's a large improvement over `"regexp"`, but it would still be best to use [`"term"`](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html) where we can (because I expect it's even faster, and because it triggers only on an exact match).  The [`"match"`](https://www.elastic.co/guide/en/elasticsearch/reference/7.13/query-dsl-match-query.html) query has to first analyze the match text and then match the result of that to the text in the field from the document.

If we were use `"match"`, I suspect that we should specify several options:

- Foremost, I think we should specify the same analyzer for the query that the indexer used (i.e., `"analyzer": "path_analyzer"`).
- Out of paranoia, I would turn off synonym generation (i.e., `"auto_generate_synonyms_phrase_query" : "false"`).  I would like to think that path-analysis would implicitly disable synonyms, but it can't hurt to be explicit.
- Similarly, I would turn off fuzzy-search (i.e., `"fuzziness": "0"`, and `"fuzzy_transpositions": "false"`).  Again, I suspect that fuzziness will be disabled by default, but it won't hurt to be explicit.
- Analogously, we want all of the terms to match, so set the operator to _and_ (i.e., `"operator": "and"`).  Again, I suspect that it would work without this, but it would be best to be explicit.

Using `"match_phrase"`, it's not obvious that those options are available or needed.  For instance, by default it uses a `slop` of zero and the analyzer defaults to match the field being matched (which seems like an obvious choice...).

So, I suggest using `"term"` on the keyword fields (`"parent"` and `"run_data_parent"`) and using `"match_phrase"` on the text field (`"directory"`).

Reply:
Sorry; yes, I meant `"term"`, not `"match"`, for the keyword fields. 🤷🏻‍♂️